File: libnd4j/include/ops/declarable/platform/llamacpp/rms_norm.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// @author Adam Gibson
//

#include <ops/declarable/OpRegistrator.h>
#include <ops/declarable/PlatformHelper.h>
#include <system/platform_boilerplate.h>

#include "llamacppUtils.h"

#if HAVE_LLAMACPP

namespace sd {
namespace ops {
namespace platforms {

/**
 * RMS Normalization using GGML kernels
 *
 * RMS Norm is used extensively in LLaMA and other modern transformer architectures.
 * Formula: x / sqrt(mean(x^2) + eps) * weight
 */
static void rmsNormLlamaCpp(NDArray* input, NDArray* weight, NDArray* output, float eps) {
    llamacppUtils::GgmlContextGuard ctx(32 * 1024 * 1024);

    struct ggml_tensor* ggml_input = llamacppUtils::createGgmlTensor(ctx, input, "input");

    // Apply RMS norm
    struct ggml_tensor* ggml_output = ggml_rms_norm(ctx, ggml_input, eps);

    // Apply weight if provided
    if (weight != nullptr && !weight->isEmpty()) {
        struct ggml_tensor* ggml_weight = llamacppUtils::createGgmlTensor(ctx, weight, "weight");
        ggml_output = ggml_mul(ctx, ggml_output, ggml_weight);
    }

    ggml_set_name(ggml_output, "output");

    // Build and execute
    struct ggml_cgraph* graph = ggml_new_graph(ctx);
    ggml_build_forward_expand(graph, ggml_output);
    llamacppUtils::executeGgmlGraph(ctx, graph);

    llamacppUtils::copyGgmlToNDArray(ggml_output, output);
}

//////////////////////////////////////////////////////////////////////////
PLATFORM_IMPL(rms_norm, ENGINE_CPU) {
    auto input = INPUT_VARIABLE(0);
    auto output = OUTPUT_VARIABLE(0);

    NDArray* weight = block.width() > 1 ? INPUT_VARIABLE(1) : nullptr;

    if (input->isEmpty()) return sd::Status::OK;

    float eps = block.getTArguments()->size() > 0 ? T_ARG(0) : 1e-6f;

    rmsNormLlamaCpp(input, weight, output, eps);

    return sd::Status::OK;
}

//////////////////////////////////////////////////////////////////////////
PLATFORM_CHECK(rms_norm, ENGINE_CPU) {
    auto input = INPUT_VARIABLE(0);
    auto output = OUTPUT_VARIABLE(0);

    Requirements req("LLAMACPP RMS_NORM OP");

    req.expectTrue(block.isUseLLAMACPP(), IS_USE_LLAMACPP_MSG);

    req.expectTrue(
        makeInfoVariable(
            [input, output] {
                return llamacppUtils::isSupportedType(input->dataType()) &&
                       llamacppUtils::isSupportedType(output->dataType());
            },
            TYPECHECK_MSG),
        NO_MSG);

    req.logTheSuccess();
    return req;
}

}  // namespace platforms
}  // namespace ops
}  // namespace sd

#endif  // HAVE_LLAMACPP

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/custom/RmsNorm.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  *  See the NOTICE file distributed with this work for additional
 *  *  information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.transforms.custom;

import lombok.Getter;
import lombok.NoArgsConstructor;
import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * RMS Norm - Root Mean Square Layer Normalization
 *
 * Implements RMSNorm as used in LLaMA, Mistral, and other modern LLMs:
 *   RMSNorm(x) = x * rsqrt(mean(x^2, axis=-1) + eps) * gamma
 *
 * This is a lighter-weight alternative to LayerNorm that skips the mean-centering
 * step, only normalizing by the root mean square.
 *
 * Inputs:
 *   0: input tensor [batch, ..., features]
 *   1: gamma weights [features] (optional - scale parameter)
 *
 * Float args:
 *   0: epsilon (default 1e-5)
 *
 * Output:
 *   0: normalized tensor [batch, ..., features]
 *
 * @author Eclipse Deeplearning4j Contributors
 */
@NoArgsConstructor
public class RmsNorm extends DynamicCustomOp {

    @Getter private double epsilon = 1e-5;

    /**
     * Create RmsNorm with input and gamma (scale weights).
     */
    public RmsNorm(SameDiff sameDiff, SDVariable input, SDVariable gamma, double epsilon) {
        super(null, sameDiff, new SDVariable[]{input, gamma}, false);
        this.epsilon = epsilon;
        addTArgument(epsilon);
    }

    /**
     * Create RmsNorm with input and gamma, default epsilon.
     */
    public RmsNorm(SameDiff sameDiff, SDVariable input, SDVariable gamma) {
        this(sameDiff, input, gamma, 1e-5);
    }

    /**
     * Create RmsNorm with input only (no gamma).
     */
    public RmsNorm(SameDiff sameDiff, SDVariable input, double epsilon) {
        super(null, sameDiff, new SDVariable[]{input}, false);
        this.epsilon = epsilon;
        addTArgument(epsilon);
    }

    /**
     * Create RmsNorm for INDArray execution.
     */
    public RmsNorm(INDArray input, INDArray gamma, INDArray output, double epsilon) {
        super(gamma != null ? new INDArray[]{input, gamma} : new INDArray[]{input},
                output != null ? new INDArray[]{output} : null);
        this.epsilon = epsilon;
        addTArgument(epsilon);
    }

    public RmsNorm(INDArray input, INDArray gamma) {
        this(input, gamma, null, 1e-5);
    }

    /**
     * Create RmsNorm for INDArray execution with explicit epsilon.
     */
    public RmsNorm(INDArray input, INDArray gamma, double epsilon) {
        this(input, gamma, null, epsilon);
    }

    /**
     * Create RmsNorm for INDArray execution without gamma.
     */
    public RmsNorm(INDArray input, double epsilon) {
        this(input, null, null, epsilon);
    }

    /**
     * Create RmsNorm for INDArray execution without gamma, default epsilon.
     */
    public RmsNorm(INDArray input) {
        this(input, null, null, 1e-5);
    }

    @Override
    public void configureFromArguments() {
        super.configureFromArguments();
        if (!tArguments.isEmpty()) {
            this.epsilon = tArguments.get(0);
        }
    }

    @Override
    public String opName() {
        return "rms_norm";
    }

    @Override
    public List<SDVariable> doDiff(List<SDVariable> gradients) {
        return Arrays.asList(new RmsNormBp(sameDiff, arg(0), gradients.get(0), epsilon).outputVariables());
    }

    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> inputDataTypes) {
        return Collections.singletonList(inputDataTypes.get(0));
    }

    @Override
    public int getNumOutputs() {
        return 1;
    }
}
